add base protocol upgrade spec (#874)
vegaprotocol/specs: protocol/0070-PLUP_protocol_upgrades.md

Feature name: protocol-upgrades
Start date: 2022-02-09

## Summary

### Current state of upgrading the vega network

As of today, upgrading the protocol is near impossible when a major changes to the step are possible without proceeding with an LNL restore.
Using LNL have the following unconvenient side effect:
- A synchronous restart is required
- All node need to be restarted in a very short time so all state can be restore from ethereum, and the network can start properly with a checkpoint.

### How different protocol proceed

> Other protocol e.g Ethereum proceed to updates in an asynchronous manner. Usually a new version of the protocol is made available, an hardcoded block height is set

Reviewer: ```suggestion
Other protocol e.g Ethereum carry out updates in an asynchronous manner. Usually a new version of the protocol is made available, a hardcoded block height is set
```

> at which a new code path will be enabled. If enough nodes runner have deployed the updated code, then the network will continue with the new code path. Others nodes

Reviewer: ```suggestion
at which a new code path will be enabled. If enough node runners have deployed the updated code, then the network will continue with the new code path. Others nodes
```

which haven't updated will then fork the blockchain.

### Prior example of updgrading the nodes asynchronously

Back in December 2020, vega proceeded to a LNL restore. Unfortunately, a bug in the code prevented the dispatch of the network parameters after the restore. This left

> the network in a semi invalid state where the network parameters where defined by a the ones from the genesis block instead of the one from the checkpoint from the previous network.

Reviewer: ```suggestion
the network in a semi invalid state where the network parameters were defined by the ones from the genesis block instead of the one from the checkpoint from the previous network.
```

The solution employed at the time was to:
- implement a hotfix

> - keep the hotfix behing a guard until a given time, once the time is reachd network parameters would be dispatched.

Reviewer: ```suggestion
- keep the hotfix behind a guard until a given time, once the time is reached network parameters would be dispatched.
```

- distribute the code to the validator so they can test it.
- then decide of an actual date which would give a week for the validator to asynchronously update their node.

This upgrade went quite smoothly without any incident, although it was possible as the fix was located in a single place, and wasn't impacting much of the state

> (meaning no changes in the protobug were required for example.)

Reviewer: ```suggestion
(meaning no changes in the protobuf files were required for example.)
```

## Asynchronous upgrades mecanism

One of the main challenge with upgrading the network both asynchronously and without a restart / downtime is

> the management of the state, and possible uncompatible changes between the state of two version of the protocol.

Reviewer: ```suggestion
the management of the state, and possible incompatible changes between the state of two version of the protocol.
```

> In order to prevent any state incompatiblity introduce a new framework to propose upgrades of the protocol, this will allow the node to be stopped,

Reviewer: ```suggestion
In order to prevent any state incompatibility, this specification introduces a new framework to propose upgrades of the protocol, which will allow the node to be stopped,
```

> and restart from the block it stopped using the snapshots or to replay the whole chain. When being restarted

Reviewer: ```suggestion
and restart from the block it stopped at using the snapshots or by replaying the whole chain. When being restarted
```

the new node will contain both the previous version of the protocol in use, but also a new version that the

> network should start using at a given block. Upong restart, the node will try to send a transaction to

Reviewer: ```suggestion
network should start using at a given block. Upon restart, the node will try to send a transaction to
```

indicate the rest of the network that it's ready to switch to the new version of the protocol. When the block
defined for the upgrade is reached and if enough nodes have send the transaction notifying of their intent of
migrating to the new version of the protocol, then the upgrade is applied. If not enough transaction have been
sent by the validators, the protocol will keep running the current version.

The existing network parameter `"validators.vote.required"` is used to defined the amount of require validators
to proceed with the update.

> The upgrade requires to be done in an atomic way. To do so, the core node will need to have been compiler with

Reviewer: ```suggestion
The upgrade **must** be done in an atomic way. To do so, the core node will need to have been compiled with
```

the current and next version of the protocol. Once the block height for the upgrade is reached the currently
running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the

> whole state will be serialized, all the engine shut down, and the new version of the protocol will be

Reviewer: ```suggestion
whole state will be serialised, all of the engines will shut down, and the new version of the protocol will be
```

> instantiated, and will have to load the state previously serialized by the old version. The new version of

Reviewer: ```suggestion
instantiated, and will have to load the state previously serialised by the old version. The new version of
```

> the protocol will be in charged to adapt / convert the previous state in order to make it compatible with

Reviewer: ```suggestion
the protocol will be then to convert the previous state in order to make it compatible with
```

any changes in the protocol.

> The most main difficulty here will be the implementation of the serialisation of the whole state of the node.

Reviewer: ```suggestion
The main difficulty here will be the implementation of the serialisation of the whole state of the node.
```

> Fortunately we already have such feature which should work out of the box with minimal changes: the snapshots.

Reviewer: ```suggestion
Fortunately we already have such feature which should work out of the box with minimal changes: [snapshots](../non-protocol-specs/0009-NP-SNAP-snapshots.md).
```

> ## Versionning of the protocol

Reviewer: ```suggestion
## Versioning of the protocol
```

From now on we will need to version the protocol, this can be the current version of the node, and the previous
version. This should probably use a versioning system like semantic versioning (this will be useful in a later
part of this document).

## Types of upgrade

Not all upgrades of the protocol may have the same impact to different kind of actors, or even the same urgency.
e.g: a patch for a critical issue, or major changes in the protocol, or simple patch.

Each of those may require the attention of different actors in the network, and different process to be accepted.

For the rest of this assume that upgrade are using a semantic version, where for a given M.m.P version which changes either of:
- M is a major upgrade
- m is a minor upgrade
- P is a patch upgrade.

Each of those upgrade will go through a different process.

### Patch upgrade

> This upgrade should represent very small change in the protocol, eventually a bug fix, or QOL improvement.

Reviewer: ```suggestion
This upgrade should represent very small change in the protocol, for example a bug fix or a quality-of-life improvement.
```

> These upgrade require no involvement from any actors of the network apart from the validator to update their node

Reviewer: ```suggestion
These upgrade require no involvement from any actors of the network apart from the validators updating their nodes
```

with a new supported version of the protocol.

### Minor upgrade

> These are upgrade which brings more features to the protocol, eventual fixes as well, and changes in the state

Reviewer: ```suggestion
These are upgrade which bring more features to the protocol, or larger fixes, and changes in the state
```

> produce by the protocol. They may require more works from the validators and will not be applied automatically.

Reviewer: ```suggestion
produced by the protocol. They may require more work from the validators and will not be applied automatically.
```

A new command will be introduce in the vega toolchain so validator can explicitly notify the network of their
support into upgrading to the new version.

### Major upgrade

These are important changes in the protocol, possibly breaking the previously known behaviours of the protocol

> and bringing also new important feature. Due to the nature of the upgrade, a governance vote will be required

Reviewer: ```suggestion
but also implementing new important features. Due to the nature of the upgrade, a governance vote will be required
```

so token holders show of their support for the upgrade.

> > note: the major upgrade will need refinment, and are not planned at first.

Reviewer: ```suggestion
> note: the definition of a major upgrade will need refinement, and the current version of the specification does not cover it.
```

## Framework / data structures

As protobuf:
```
message Upgrade {
	enum Type {
		Patch = 1;
		Minor = 2;
		Major = 3;
	}
	string new_version = 1; // the new version proposed for the upgrade. in semver
	uint64 block_height = 2; // the block at which the upgrade will be executed.
	string description = 3; // a description for this upgrade
}

// a message sent by the node to notify that a node is ready to do an upgrade
message UpgradeAccepted {
	string version = 1;
}
```

## Acceptance criteria
- [ ] A new patch upgrade is made available

>   - [ ] Enough validator have upgraded their node

Reviewer: ```suggestion
  - [ ] Enough validators have upgraded their node
```

Reviewer: Also this one could be clearer what 'enough' means (e.g. 2/3+1 have signalled with a X event)

Start date: 2022-02-09

## Summary

### Current state of upgrading the vega network

As of today, upgrading the protocol is near impossible when a major changes to the step are possible without proceeding with a [Limited Network Life checkpoint restore](./non-protocol-specs/0005-NP-LIMN-limited_network_life.md). This functionality has the following significant issues:
- A synchronous restart is required
- All node need to be restarted in a very short time so all state can be restore from ethereum, and the network can start properly with a checkpoint.

Limited Network Life is not the end goal. This spec outlines how the protocol evolves from LNL checkpoints to rolling software updates, controlled by a reasonable set of governance and user controls.

### How other protocols proceed

Other protocol e.g Ethereum carry out updates in an asynchronous manner. Usually a new version of the protocol is made available, a hardcoded block height is set
at which a new code path will be enabled. If enough node runners have deployed the updated code, then the network will continue with the new code path. Others nodes
which haven't updated will then fork the blockchain.

> 

Reviewer 2: Eth may not be the best example, as it has a different behaviour (mainly, if their upgrade causes disagreement., they forc; if ours causes disagreement, the chain stops until sufficient weight supports either version. Maybe refer to how cosmos does it (https://docs.cosmos.network/master/core/upgrade.html) ?

Author: Yes, I was using them as an example of keeping the chain running while processing upgrade, not specifically how they do them. The cosmos way is how we will do our upgrade indeed.

### Prior example of updgrading the nodes asynchronously

> Back in December 2020, vega proceeded to a LNL restore. Unfortunately, a bug in the code prevented the dispatch of the network parameters after the restore. This left

Reviewer 3: I think it was 2021, not 2020.

## Asynchronous upgrades mechanism

One of the main challenge with upgrading the network both asynchronously and without a restart / downtime is
the management of the state, and possible incompatible changes between the state of two version of the protocol.

In order to prevent any state incompatibility, this specification introduces a new framework to propose upgrades of the protocol, which will allow the node to be stopped,
and restart from the block it stopped at using the snapshots or by replaying the whole chain. When being restarted
the new node will contain both the previous version of the protocol in use, but also a new version that the
network should start using at a given block. Upon restart, the node will try to send a transaction to
indicate the rest of the network that it's ready to switch to the new version of the protocol. When the block
defined for the upgrade is reached and if enough nodes have send the transaction notifying of their intent of
migrating to the new version of the protocol, then the upgrade is applied. If not enough transaction have been
sent by the validators, the protocol will keep running the current version.

> 

Reviewer 2: Those transactions are sent through the chain as well to make sure we have consensus ? And, what is enough ? 2/3 of the validator weight (which allows to run, but no longer for failures) or more ?

Author: Yea that should be 2/3 of the validators weight I suppose.

The existing network parameter `"validators.vote.required"` is used to defined the amount of require validators
to proceed with the update.

The upgrade **must** be done in an atomic way. To do so, the core node will need to have been compiled with
the current and next version of the protocol. Once the block height for the upgrade is reached the currently
running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the

> whole state will be serialised, all of the engines will shut down, and the new version of the protocol will be

Reviewer 4: This is just an observation really. If there is a bug in the code that serialising the whole state of vega (snapshots), for example a vital peice of state is not being captured, then we're screwed right? Because we won't be able to migrate that value into the new vega version.

Unless we replay the whole chain with a new version that has *just* the snapshot fix in. Maybe.

Author: yes that's correct, we would miss the data, an replaying the wholke chain would be needed, and applying the fix only when we reach the point in the chain where the upgrade is supposed to be applied, then the state would be save from this point.

instantiated, and will have to load the state previously serialised by the old version. The new version of
the protocol will be then to convert the previous state in order to make it compatible with
any changes in the protocol.

The main difficulty here will be the implementation of the serialisation of the whole state of the node.
Fortunately we already have such feature which should work out of the box with minimal changes: [snapshots](../non-protocol-specs/0009-NP-SNAP-snapshots.md).

## Versioning of the protocol

> 

Reviewer 2: Since we're (at least in theory) not the only provider of code, the version should probably also contain a hash to make sure it's uniquely defined. Otherwise, it would be possible that validators think they agree on the version (as the numbers match), but run code from different providers. Another way would be that Vega still signs the code, and thus gives their version a unique vendor ID (I'm not sure how far we do that now).
The latter might be a good idea anyhow, as that would create a well defined authentication of update code, and we can implement 4-eye security and such.

Author: Well I think that ideally we don't want to vendor lock vega as the only provider of the code. We version the protocol here, and fixes to the protocol, so I would expect that version N of the protocol should behave the same accros different implementation. The vega node itself though, is version separately.

Reviewer 5: Need a way to inform the validators that they cannot upgrade from version 1 to 3 and they need to update to version 2 first then 3. This also needs to account for less simple cases where version numbers are not sequential (i.e. semver + serial number).

> 

Reviewer 2: Another point that should be clarified with versioning is the downgrade option - is it ok to downgrade if a new version has a problem, or should someone in that case provide a new version number with the older code (i.e., the code would refuse to update to a lower version number) ? I suppose given that an update code has both versions of the protocol, a downgrade would require a new code blob from us ?
(It might be good if the validators could switch back if needed; however, this would mean that there old version can parse the serialisation of the new one, which would need a extra layer of code and not be overly feasible.

Author: Yea, we could decide to do that. That would be a massive difficulty, is something was to change so much that you cannot use it in a previous version. I would instead prefer to have to upgrade all the time, and for us to provide a fix version that the validators would use to upgrade to.

Reviewer 2: Agree here; the only issue is that if we need a fast rollback due to an issue, we still have to be provide the functionality for the rollback version (now with the higher version number) to read the serialisation of the current version, so a fast rollback isn't always possible.

Author: Well, I'd imagine a rollback would be us providing a new version, where the upgrade is the previous version and the legacy version is the upgrade, with a new layer in between to convert back the data to the previous state?

Reviewer 5: @gordsport - add risk to risk log of complex rollback if indeed possible.

@jeremyletang Can we have, in the future, an activate/deactivate process to mitigate this risk? (new spec on feature toggling?)

From now on we will need to version the protocol, this can be the current version of the node, and the previous
version. This should probably use a versioning system like semantic versioning (this will be useful in a later
part of this document).

### Edge cases
With the upgrade process outlined below, it's possible for an upgrade to be proposed and rejected by validators. This happens independently of the *protocol development* and could lead to a situation in which an upgrade is rejected, and subsequent versions need to be reworked to take this in to account.

For example:
- In a network of 3 nodes, all running v`0.1.1`
- v`0.2.0` is released with block height of 100
- 0 nodes update before block 100, so the new version is not deployed on the network
- Meanwhile v`0.3.0` is released

The network is now in a situation where `v0.3.0` expects the network to be upgrading from v`0.2.0`, while it would actually be upgrading from v`0.1.0`. This is expected, and upgrades must take this in to account. 

> 

Reviewer 2: I'm not sure that's compatible with what's described in line 53; given the code has both the old and the new version, and the code needs to be able to parse and transfer the state, jumping an update should not be possible (unless I misunderstood something). 
Maybe it would be an idea to also add a serial number to upgrades or add to every upgrade which version it expects to upgrade from(so they have to come in order), and it is our job to keep track of what the network is doing and add the right serial number/code

Author: well, a code release will be compatible only for upgrade from a version to another. Now if a version is not accepted in time by validators, then it will be for us to provide code again, with the same legacy version than in production, and change the upgrade path from 1 -> 2 to 1 -> 3 (meaning the convertion code has to be re-written etc, version 2 is discarded completely).

but also implementing new important features. Due to the nature of the upgrade, a governance vote will be required
so token holders show of their support for the upgrade.

> note: the definition of a major upgrade will need refinement, and the current version of the specification does not cover it.

## Framework / data structures

As protobuf:
```go
message Upgrade {
	enum Type {
		Patch = 1;
		Minor = 2;
		Major = 3;
	}
	string new_version = 1; // the new version proposed for the upgrade. in semver
	uint64 block_height = 2; // the block at which the upgrade will be executed.
	string description = 3; // a description for this upgrade
}

// a message sent by the node to notify that a node is ready to do an upgrade
message UpgradeAccepted {
	string version = 1;
}
```

## Acceptance criteria
### Patch version updates (`0.0.x`)
- [ ] A new patch upgrade is made available, and is successfully deployed (<a name="0070-PLUP-001" href="#0070-PLUP-001">0070-PLUP-001</a>)
  - [ ] Enough validators have upgraded their node
  - [ ] At least `validators.vote.required` message have been sent by the validators
  - [ ] The block time for the upgrade is reached

>   - [ ] The network changes to the new version of the protocol (can be checked with APIs)

Reviewer 4: Are we going to entertain the idea that if a patch does not contain changes the break consensus, then a set of validators could be running with the same minor version, but will different patch version? Or are we just enforcing that all validators need to be running identical software, even down to the patch number?

Author: Well, I'm not sure about this. I'd imagine that we would want everyone to update to the last version just in case.
Do you think it would be too much of a hassle? I imagine that a patch could actually be something that fix a potential consensus issue in a future block too.

Reviewer 4: I only asks because its something tendermint supports in that is separates overall version number from the "protocol version". The easiest and safest thing certainly would be to have everyone running identical code.

  - [ ] Enough validators have upgraded their node
  - [ ] Less then `validators.vote.required` validators sent a command to accept the upgrade
  - [ ] The block time for the upgrade is reached
  - [ ] The network does not change version (can be checked with APIs)
- [ ] A new minor upgrade is made available, but not enough validators update by the block height (<a name="0070-PLUP-005" href="#0070-PLUP-005">0070-PLUP-005</a>)
  - [ ] Not enough validators have upgraded their node
  - [ ] The block time for the upgrade is reached
  - [ ] The network does not change version (can be checked with APIs)
 
### Major version updates (`x.x.x`)
 > Note: These acceptance criteria are *incomplete*. There will likely be a governance vote from users, as well as the validator vote. For now these Acceptance Criteria test a major version in the same was a minor version upgrade.
- [ ] A new major upgrade is made available, validators vote to update, and the update is successfully deployed (<a name="0070-PLUP-006" href="#0070-PLUP-006">0070-PLUP-006</a>)
  - [ ] Enough validators have upgraded their node

>   - [ ] A governance proposal have been created to propose the upgrade

Reviewer 2: Would the governance proposal be hard-coded ? I.e., if all validators upgrade anyhow, would the code stop them ?
If it is, can the validators use the upgrade on testnet anyhow ? Do they have a way to simulate a governance proposal there easily that's compatible with the enforcement?

Author: I'm not sure we are speaking of a governance proposal here. It's more of a framework internal to the protocol, just for the validators to accept or not the release manually

Reviewer 5: Do not expect this will be used right now - still needs to be specified

  - [ ] The proposal have been accepted
  - [ ] The block time for the upgrade is reached
  - [ ] The network has switched to the new version of the protocol (can be checked with APIs)
- [ ] A new major upgrade is made available, but the validator vote does not pass (<a name="0070-PLUP-007" href="#0070-PLUP-007">0070-PLUP-007</a>)
  - [ ] Enough validators have upgraded their node
  - [ ] A governance proposal have been created to propose the upgrade
  - [ ] The proposal have been rejected

>   - [ ] The block time for the upgrade is reached

Reviewer 2: We used block height before. Both works, but needs to be clear which one we use.

Author: I don't mind as well, what's the most practical / speaking better to users?

Reviewer 5: Should we use both i.e. block height for patch and block time for major?

Do we also need an option in case we never reach the height or time due to the chain having stopped?
